### ISSUE TITLE
Does this support multiprocessing AND threads in the processes?

### ISSUE THREAD

**Issue by brianmacy** [alex-petrenko/faster-fifo#32]
<I am importing the custom pickler>
I get this AssertionError trigged a lot...

```
Exception in thread Thread-6:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "./python/G2Loader.py", line 879, in g2Thread
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    row = workQueue_.get(True, 1)
  File "faster_fifo.pyx", line 197, in faster_fifo.Queue.get
    self._target(*self._args, **self._kwargs)
  File "./python/G2Loader.py", line 879, in g2Thread
  File "faster_fifo.pyx", line 179, in faster_fifo.Queue.get_many
    row = workQueue_.get(True, 1)
  File "faster_fifo.pyx", line 197, in faster_fifo.Queue.get
  File "faster_fifo.pyx", line 215, in faster_fifo.Queue.parse_messages
  File "faster_fifo.pyx", line 179, in faster_fifo.Queue.get_many
AssertionError
  File "faster_fifo.pyx", line 215, in faster_fifo.Queue.parse_messages
AssertionError
Exception in thread Thread-16:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
Exception in thread Thread-14:
```

If I catch the assertion and allow the thread to continue I get all sort of random errors:
```
Exception in thread Thread-6:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "./python/G2Loader.py", line 879, in g2Thread
    row = workQueue_.get(True, 1)
  File "faster_fifo.pyx", line 197, in faster_fifo.Queue.get
  File "faster_fifo.pyx", line 179, in faster_fifo.Queue.get_many
  File "faster_fifo.pyx", line 212, in faster_fifo.Queue.parse_messages
  File "faster_fifo.pyx", line 64, in faster_fifo.Queue.loads
_pickle.UnpicklingError: could not find MARK
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "./python/G2Loader.py", line 879, in g2Thread
    row = workQueue_.get(True, 1)
  File "faster_fifo.pyx", line 197, in faster_fifo.Queue.get
  File "faster_fifo.pyx", line 179, in faster_fifo.Queue.get_many
  File "faster_fifo.pyx", line 212, in faster_fifo.Queue.parse_messages
  File "faster_fifo.pyx", line 64, in faster_fifo.Queue.loads
_pickle.UnpicklingError: pickle data was truncated
```

**Comment by brianmacy** [comment]
Certainly shouldn't be... basically just reading from a file, putting the string on the FIFO, and getting it on the other end.

I'll see if I can reproduce with a small test case.

**Comment by alex-petrenko** [comment]
Thank you for the update.

Could it be some actual asserts from the main pyx file, like https://github.com/alex-petrenko/faster-fifo/blob/cf03f788944cd70f4c5cd18693af3224967f797b/faster_fifo.pyx#L177 ?

I wonder if there is a way to know which particular assert is triggered in Cython. Maybe add text messages to them?
like `assert <smth>, "message"`

Does your problem reproduce with large strings or relatively small strings are enough? Is there a chance you're overflowing the buffer size? I think I have tests for that but this is a place with the highest probablility for bugs since people must face this quite rarely

**Comment by brianmacy** [comment]
Overflowing the buffer is absolutely likely, but not with a single message.  It runs fine for like 10s of thousands of strings and then chokes.  I was creating a Queue of max size of 4GB (4*1024*1024*1024) and the largest string in the first 1M messages is 11.3MB and the largest in the first 10M messages is 39MB.  I don't think I even got through the first 100k messages (250KB).

**Comment by alex-petrenko** [comment]
Hmm... This is a nice stress test. I might reveal some bug the other tests couldn't catch. In theory there should be no reason for this not to work, so must be a bug.

Some kind of synthetic test for this would be an amazing contribution! :)

**Comment by brianmacy** [comment]
[ff_test.tgz](https://github.com/alex-petrenko/faster-fifo/files/9140585/ff_test.tgz)

$ python3 --version
Python 3.8.10
$ cat /etc/issue
Ubuntu 20.04.4 LTS \n \l

The box has 32 physical cores (64 logical) and 256GB RAM.  A couple of things to note.

1. If you change the subprocess to be single-threaded it works fine.  Check change the "range" from  8 to 128 and have it spawn reader(...) instead of reader_process(...).

2. If you do step (1) BUT then change it to spawn 128 threads by changing Process to threading.Thread, it will fail.  Essentially confirming my suspicion that it is somehow protecting cross process but not cross thread.

I notice you initialize the mutex with a default pthread_mutexattr_t structure.  Note that there are conditions with the PTHREAD_MUTEX_DEFAULT that are undefined.  You might want to make sure PTHREAD_MUTEX_NORMAL is used.  Also, the pthread_mutex_lock return code is not actually checked.

**Comment by brianmacy** [comment]
I am fairly certain that the issue is in faster_fifo.pyx

    def reallocate_msg_buffer(self, new_size):
        new_size = max(INITIAL_RECV_BUFFER_SIZE, new_size)
        self.message_buffer = (ctypes.c_ubyte * new_size)()
        self.message_buffer_memview = memoryview(self.message_buffer)

Those buffers are global to the queue object but are not protected by a mutex.  When you call queue_get you are releasing the GIL so those shared buffers are used now in other threads but you have not finished with parse_messages yet.

If there is a way to declare it as thread_local then it probably will resolve it.  From a quick google, this is going to need another solution.

**Comment by alex-petrenko** [comment]
@brianmacy thank you for looking into this!

Looking back, indeed all of my applications just used multiprocessing which guarantees that this recv buffer is completely separate in each case. Otherwise, there's indeed no thread safety. I overlooked this.

I like using thread local storage as a solution. Actually went ahead and measured performance impact before and after your patch, and results seem almost indistinguishable.

```
Before patch:
# [2022-07-20 19:15:17,836][2343248] Configuration (1, 1, 200000), timing [ff: 1.04s, ff_many: 1.10s, mp.queue: 2.25s]
# [2022-07-20 19:15:17,836][2343248] Configuration (1, 10, 200000), timing [ff: 1.65s, ff_many: 1.78s, mp.queue: 3.51s]
# [2022-07-20 19:15:17,836][2343248] Configuration (10, 1, 100000), timing [ff: 14.16s, ff_many: 1.03s, mp.queue: 12.07s]
# [2022-07-20 19:15:17,836][2343248] Configuration (3, 20, 100000), timing [ff: 3.18s, ff_many: 2.29s, mp.queue: 7.63s]
# [2022-07-20 19:15:17,836][2343248] Configuration (20, 3, 50000), timing [ff: 15.29s, ff_many: 0.69s, mp.queue: 17.45s]
# [2022-07-20 19:15:17,836][2343248] Configuration (20, 20, 50000), timing [ff: 1.43s, ff_many: 3.97s, mp.queue: 26.52s]

# after patch:
# [2022-07-20 19:19:02,880][2344018] Configuration (1, 1, 200000), timing [ff: 1.05s, ff_many: 1.13s, mp.queue: 2.24s]
# [2022-07-20 19:19:02,880][2344018] Configuration (1, 10, 200000), timing [ff: 1.70s, ff_many: 1.76s, mp.queue: 3.54s]
# [2022-07-20 19:19:02,880][2344018] Configuration (10, 1, 100000), timing [ff: 14.34s, ff_many: 0.96s, mp.queue: 12.16s]
# [2022-07-20 19:19:02,880][2344018] Configuration (3, 20, 100000), timing [ff: 3.15s, ff_many: 2.30s, mp.queue: 7.87s]
# [2022-07-20 19:19:02,880][2344018] Configuration (20, 3, 50000), timing [ff: 15.32s, ff_many: 0.67s, mp.queue: 17.48s]
# [2022-07-20 19:19:02,880][2344018] Configuration (20, 20, 50000), timing [ff: 1.45s, ff_many: 3.97s, mp.queue: 26.99s]
```

I'll also add fprintf to asserts, great suggestion! Probably going to add some threading tests too.

> Doing some testing. One of the funny things about this is that each buffer grows to the biggest message size and stays there. They all grow at the same time too as each thread/process tends to get a chance at the record that causes the failure to get based on size... so they all resize.

I'd say this works as intended then? :)

> For systems with highly varying sizes, it probably makes sense to size back down or to allocate a one-off buffer once it gets over a certain threshold. It would also be nice if there was a way to resize the shared memory and do the queue to be number of items based, instead of bytes.

Yes, nice feature suggestions. I guess my philosophy was to optimize for speed at the expense of memory + trying to keep it simple. If you're transfering truly big objects between processes maybe it makes sense to allocate some shared memory storage and just send pointers to this storage? This will save a ton of CPU time on pickling/unpickling.

**Comment by alex-petrenko** [comment]
Released PyPI version 1.4.1 with pretty much your fix as-is. Also improved error messages.
Thanks a lot for finding this!

**Comment by brianmacy** [comment]
Great to hear that it worked for you too.

I noodled turning it into a shared buffer of pointers to messages.  Not sure what shm alloc performance is in that case BUT it would make the contention around the queue tiny and eliminate some extra copying.  All you would do in a put would be to see if a "slot" was open, allocate a shm buffer, copy the message, and put the pointer in the array.   On a get you would just check if a slot was available and return it... in neither case would you ever have to worry about a buffer being big enough.

Then I started to wonder if that is what Multiprocessing.Queue does but I haven't had the desire to go look.
